src/algo/traversal/dfs/mod.rs

```diff
@@ -29,57 +27,72 @@ pub enum EdgeType {
     CrossEdge,
 }
 
-pub enum Event<'a, G: NodeIdMapProvider> {
-    Begin(&'a State<G>, NodeId),
-    Discover(&'a State<G>, NodeId),
-    Finish(&'a State<G>, NodeId),
-    End(&'a State<G>),
-    VisitEdge(&'a State<G>, EdgeType, NodeId, NodeId),
+pub enum Event<'a, I: IdMap> {
+    Begin(&'a State<I>, NodeId),
+    Discover(&'a State<I>, NodeId),
+    Finish(&'a State<I>, NodeId),
+    End(&'a State<I>),
+    VisitEdge(&'a State<I>, EdgeType, NodeId, NodeId),
 }
 
-pub struct State<G: NodeIdMapProvider> {
-    pub id_map: <G as NodeIdMapProvider>::NodeIdMap,
+pub struct State<I: IdMap> {
+    pub id_map: I,
 
     pub time: usize,
     pub discover: Vec<usize>,
     pub finished: Vec<usize>,
     pub parent: Vec<usize>,
-
-    pub(crate) phantom_g: PhantomData<G>,
 }
 
-pub struct DFS<'a, G>
+pub struct DFS<'a, G, I = DefaultIdMap>
 where
-    G: NodeProvider + NodeIdMapProvider,
+    G: NodeProvider,
+    I: IdMap,
 {
     graph: &'a G,
     starting_nodes: Vec<NodeId>,
-    state: State<G>,
+    state: State<I>,
 }
 
 impl<'a, G> DFS<'a, G>
 where
-    G: NodeProvider + NodeIdMapProvider,
+    G: NodeProvider,
 {
     pub fn init(graph: &'a G) -> Self {
-        Self::with_starters(graph, None.into_iter())
+        Self::with_ids_and_starters(graph, DefaultIdMap::new(graph), None.into_iter())
```

Comment:
Shouldn't `DefaultIdMap` be `RudyIdMap`? Considering it's both faster and uses less memory, I thought we arrived at the conclusion to use it as the default id map.

Reply:
Got it. I've made `RudyIdMap` the default id map.